<p></p>
<p>P.R.A.T. <strong>(Primer Checker Results Analysis Tool)</strong> for <strong>SARS-CoV-2</strong> is a tool to help you
analyse the <strong><a href="https://www.gisaid.org"  tartget="_blank">GISAID’s</a> Primer Checker</strong> results about <strong>SARS-CoV-2</strong>.</p>
<p>It uses as input a <strong>Primer Checker results file</strong> containing
mutations for forward/reverse primer and probe. One must also input the total
number of SARS-CoV-2 variants Primer Checker used.</p>
<p>The results is a series of tables and graphs that can be used in publications
presneting the information from the Primer Checker about SARS-CoV-2.</p>
<hr/>
<h3>Start Analysis</h3>
<p>You start your analysis by selecting the <strong>Start tab</strong></p>
Here the user  
<ol>
  <li>must upload the result file obtained from GISAID. The provided file, already obtained from GISAID can be used as an example. </li>
<li>must specify the number of the variants. (Unfortunately, the file itself does not contain this information). </li>
<li>may, optionally, specify a date range. </li>
When field are filled, the analysis takes place automatically. 
At its end, the <strong>Results</strong> tab can be selected. 
</ol>
<h3>Reviewing the results</h3>
By selecting the Results tab the user moves to the Results of the Analysis of the input file he has provided. 
The results are split in 5 tabs: 
<ul>
  <li>Original Data<p>This is an overview of the input GISAID file. All the alignments of the query file used in GISAID are shown here (hence “Original Data”). Here the user may see all alignments of the query sequences with the sequences deposited in GISAID along with any mismatches. For each alignment there is more information that can be searched through the filtering fields at the top of the list. Additionally, the results may be sorted according to any column. </p><img src="results_original_data.jpg"></li>
  <li>Filter parameters<p>The filter parameters allow for further customization of the analysis results. For the time being the user may specify a more restrictive Date Range or restrict the results to one or more specific countries. The results shown in the ‘Original Data’ column won’t change but the results analysed in the following tabs will. </p></li>
  <li>Statistics 
    <ul>
    <li>table 1: <p>the number of cumulative mutations, as a total and as a percentage. </p></li>
    <li>table 2, Position of Single Mutations: The position of each mutation and the mutated nucleotide are shown. </p></li>
    <li>table 3, Detailed Single Mutations for forward primer type: <p>the cumulative number of mutations in each position and for each nucleotide for the forward primer. </p></li>
    <li>table 4, Detailed Single Mutations for rev primer type:<p>Same as table 3, for the reverse primer.</p></li>
    </ul>
  </li>
  <li>Graph for fwd primer <p>The graph for the forward primer can be customized. More specifically, one can customize: </p>
  <ul>
    <li>each nucleotide color </li>
    <li>the Main graph title and the 2 axis labels </li>
    <li>the Cut-off Percent value. Only labels above the cutoff value are shown in the graph.</li>
  </ul>
  After customization, the user may download the graph in a high-quality pdf,jpg, png or tiff format.</li>
</ul>

<!-- <p>From the following dropdown menu you can select your theme:</p> -->
